tests/unittest_inference.py

```diff
@@ -2642,11 +2642,11 @@ def true_value():
         klass = module["Class"]
         self.assertTrue(klass.bool_value())
         dict_comp = next(module["dict_comp"].infer())
-        self.assertEqual(dict_comp, util.Uninferable)
+        assert isinstance(dict_comp, nodes.DictComp)
         set_comp = next(module["set_comp"].infer())
-        self.assertEqual(set_comp, util.Uninferable)
+        assert isinstance(set_comp, nodes.SetComp)
         list_comp = next(module["list_comp"].infer())
-        self.assertEqual(list_comp, util.Uninferable)
+        assert isinstance(list_comp, nodes.ListComp)
```

Comment:
Hm, yeah I didn't think of it like that. That would decouple the value of `lst` from the actual `ast` tree though.

For example, we would then have a `ListComp` node in `nodes.Module.body` but a `List` node for `lst`? If we create that `List` when calling `_infer` on `ListComp` what would then be the `lineno` etc.?

Wouldn't it be better to say that `lst` is a `ListComp` and let downstream users decide if they want to handle `ListComp` and `List` similarly? Like we often do with `FunctionDef` and `AsyncFunctionDef`.

Reply:
> That would decouple the value of `lst` from the actual `ast` tree though.

Yeah, it's an inference **result** after all.

> If we create that `List` when calling `_infer` on `ListComp` what would then be the `lineno` etc.?

`None`? Both `lineno` and `col_offset` attributes are typed as optional.
Or the same one as the original `ListComp`.

> Wouldn't it be better to say that `lst` is a `ListComp` and let downstream users decide if they want to handle `ListComp` and `List` similarly? Like we often do with `FunctionDef` and `AsyncFunctionDef`.

The downstream user still has full control over it. If he once the "raw" data, he can just use the node itself. No need to call `infer` for that. Regarding `FunctionDef`, the infer method can actually return a `Property` if the function is decorated. Similarly, inferring a `DictUnpack` will return a single dict or a comparison just a bool constant
```py
{1: 'A', **{2: 'B'}}
True == True
```

Reply:
I guess we should probably handle this comprehension per comprehension then. Instead of one larger PR.

Reply:
> I guess we should probably handle this comprehension per comprehension then. Instead of one larger PR.

Yeah. There are multiple parts to be aware of: evaluating the iterator, matching the loop var(s) with optional sequence unpacking, checking any if statements and doing late assignments via `:=`, evaluating the loop expression. Likely I'm forgetting something.

```py
[a + c for a, b in func() if (c := b * 2)]
```

If you, or someone else wants to go down that route, it's probably best to start with a fairly limited set constraints and only one comprehension type. Once that is down extending it should be fairly easy in comparison.